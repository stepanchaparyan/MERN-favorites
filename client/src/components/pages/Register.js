import React, { useContext, useState, useEffect } from 'react';
import { Link } from 'react-router-dom';
import AuthContext from '../../context/authContext/authContext';
import {
  Container,
  Title,
  Form,
  Input,
  Errors,
  QuestionText,
  ErrorButton
} from './RegisterStyled';

const Register = props => {
  const { register, isAuthencated, error, clearErrors, setError } = useContext(
    AuthContext
  );
  useEffect(() => {
    if (isAuthencated) {
      props.history.push('/');
    }
  }, [isAuthencated, props.history]);

  const [user, setUser] = useState({
    name: '',
    email: '',
    password: '',
    password2: ''
  });
  const { name, email, password, password2 } = user;
  const onchange = e => {
    setUser({ ...user, [e.target.name]: e.target.value });
    if (error !== null) {
      clearErrors();
    }
  };
  const onsubmit = e => {
    e.preventDefault();
    if (password !== password2) {
      setError('Password does not match');
    } else {
      register({
        name,
        email,
        password
      });
    }
  };
  return (
    <Container>
      <Title>Sign Up</Title>
      <Form onSubmit={onsubmit}>
        <Input
          type='text'
          name='name'
          placeholder='Name'
          value={name}
          onChange={onchange}
          required
        />
        <Input
          type='email'
          name='email'
          placeholder='Email'
          value={email}
          onChange={onchange}
          required
        />
        <Input
          type='password'
          name='password'
          placeholder='Password'
          value={password}
          onChange={onchange}
          required
        />
        <Input
          type='password'
          name='password2'
          placeholder='Confirm Password'
          value={password2}
          onChange={onchange}
          required
        />
        <Input type='submit' value='Sing Up' />
      </Form>
      {error !== null && (
        <Errors>
          {!Array.isArray(error) ? (
            <ErrorButton type='button' onClick={() => clearErrors()}>
              {error}
            </ErrorButton>
          ) : (
            error.map(err => (
              <ErrorButton
                key={err.msg}
                type='button'
                onClick={() => clearErrors()}
              >
                {err.msg}
              </ErrorButton>
            ))
          )}
        </Errors>
      )}
      <QuestionText>
        Already have an accout? &nbsp;<Link to='/login'>Sign In</Link>
      </QuestionText>
    </Container>
  );
};

export default Register;
