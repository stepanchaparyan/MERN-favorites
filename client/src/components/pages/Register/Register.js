import React, { useContext, useState, useEffect } from 'react';
import { useIntl } from 'react-intl';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import AuthContext from '../../../context/authContext/authContext';
import {
  Container,
  Title,
  Form,
  Input,
  Errors,
  QuestionText,
  ErrorButton,
  RegisterButton
} from './RegisterStyled';
import localization from './localization';
import { FORM, LINK, ERRORS } from '../../../constants';

const { INPUT } = FORM;

const Register = props => {
  const { register, isAuthencated, error, clearErrors, setError } = useContext(AuthContext);
  const { formatMessage } = useIntl();

  useEffect(() => {
    if (isAuthencated) {
      props.history.push('/');
    }
  }, [isAuthencated, props.history]);

  const [user, setUser] = useState({
    name: '',
    email: '',
    password: '',
    password2: ''
  });
  const { name, email, password, password2 } = user;
  const onchange = e => {
    setUser({ ...user, [e.target.name]: e.target.value });
    if (error !== null) {
      clearErrors();
    }
  };
  const onsubmit = e => {
    e.preventDefault();
    if (password !== password2) {
      setError(ERRORS.TEXT.PASSWORD_MATCH);
    } else {
      register({
        name,
        email,
        password
      });
    }
  };
  return (
    <Container>
      <Title>{formatMessage(localization.signup)}</Title>
      <Form onSubmit={onsubmit}>
        <Input
          type={INPUT.TYPE.TEXT}
          name={INPUT.NAME.NAME}
          placeholder={formatMessage(localization.name)}
          value={name}
          onChange={onchange}
          required
        />
        <Input
          type={INPUT.TYPE.EMAIL}
          name={INPUT.NAME.EMAIL}
          placeholder={formatMessage(localization.email)}
          value={email}
          onChange={onchange}
          required
        />
        <Input
          type={INPUT.TYPE.PASSWORD}
          name={INPUT.NAME.PASSWORD}
          placeholder={formatMessage(localization.password)}
          value={password}
          onChange={onchange}
          required
        />
        <Input
          type={INPUT.TYPE.PASSWORD}
          name={INPUT.NAME.PASSWORD_2}
          placeholder={formatMessage(localization.confirmPassword)}
          value={password2}
          onChange={onchange}
          required
        />
        <RegisterButton>{formatMessage(localization.signup)}</RegisterButton>
      </Form>
      {error !== null && (
        <Errors>
          {!Array.isArray(error) ? (
            <ErrorButton type={INPUT.TYPE.BUTTON} onClick={() => clearErrors()}>
              {error}
            </ErrorButton>
          ) : (
            error.map(err => (
              <ErrorButton key={err.msg} type={INPUT.TYPE.BUTTON} onClick={() => clearErrors()}>
                {err.msg}
              </ErrorButton>
            ))
          )}
        </Errors>
      )}
      <QuestionText>
        {formatMessage(localization.alreadyHaveAnAccout)} &nbsp;
        <Link to={LINK.TO.LOGIN}>{formatMessage(localization.signin)}</Link>
      </QuestionText>
    </Container>
  );
};

Register.propTypes = {
  history: PropTypes.object
};

export default Register;
