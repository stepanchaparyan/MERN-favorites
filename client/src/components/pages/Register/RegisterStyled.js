import styled from 'styled-components';
import { tabletUp } from '../../../styles/mediaQueries/mixins';

export const Container = styled.div`
  display: flex;
  justify-content: center;
  flex-direction: column;
  color: ${props => props.theme.cadetblue};
`;

export const Title = styled.div`
  display: flex;
  justify-content: center;
  padding: 64px 32px 32px;
  font-weight: 600;
  font-size: 30px;
`;

export const Form = styled.form`
  display: flex;
  flex-direction: column;
  align-items: center;
`;

export const Input = styled.input`
  border: 0px;
  border-bottom: 1px solid ${props => props.theme.cadetblue};
  width: 50%;
  padding: 7px 10px;
  margin: 2px;
  outline: none;
  ${tabletUp`
    width: 30%;
  `};

  &[type='submit'] {
    width: 40%;
    border: 0;
    background-color: ${props => props.theme.cadetblue};
    font-weight: 600;
    font-size: 16px;
    margin-top: 20px;
    border-radius: 4px;
    cursor: pointer;
    ${tabletUp`
      width: 20%;
    `};
  }
`;

export const Errors = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  margin: 10px;
`;

export const QuestionText = styled.div`
  display: flex;
  align-self: center;
  margin: 10px;
  & > a {
    text-decoration: none;
    color: black;
    font-weight: 600;
  }
`;

export const ErrorButton = styled.button`
  color: red;
  background-color: transparent;
  border: none;
  outline: none;
  cursor: pointer;
`;
