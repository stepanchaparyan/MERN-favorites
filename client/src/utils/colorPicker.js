export const colorPicker = category => {
  let color;
  switch (category) {
    case 'Film':
      color = 'red';
      break;
    case 'Music':
      color = 'blue';
      break;
    case 'Books':
      color = 'limeGreen';
      break;
    case 'Other':
      color = 'orange';
      break;
    default:
      color = 'black';
  }
  return color;
};
