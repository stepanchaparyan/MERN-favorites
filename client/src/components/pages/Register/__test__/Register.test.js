import React, { useContext } from 'react';
import { renderSnapshot } from '../../../../utils/tests';
import { shallow } from 'enzyme';
import 'jest-styled-components';
import Register from '../Register.js';
import { Input, Form, ErrorButton } from '../RegisterStyled';

jest.mock('react-intl', () => ({
  useIntl: jest.fn(() => ({ formatMessage: m => m.defaultMessage })),
  defineMessages: i => i
}));

jest.mock('react', () => {
  const ActualReact = jest.requireActual('react');
  return {
    ...ActualReact,
    useEffect: jest.fn().mockImplementation(f => f()),
    useContext: jest.fn().mockImplementation(() => ({
      isAuthencated: false,
      error: 'a',
      clearErrors: jest.fn(),
      register: jest.fn()
    }))
  };
});

describe('Register component test with Enzyme', () => {
  test('renders snapshot', () => {
    renderSnapshot(Register);
  });

  test('should click on Input elements', () => {
    const component = shallow(<Register />);
    const inputs = component.find(Input);
    inputs.forEach(input => {
      input.simulate('change', { target: { value: 'target' } });
    });
  });

  test('should click on Form elements', () => {
    const component = shallow(<Register />);
    const form = component.find(Form);
    const preventDefault = jest.fn();
    form.simulate('submit', { preventDefault });
    expect(preventDefault).toHaveBeenCalled();
  });

  test('should click on errorButton elements', () => {
    const component = shallow(<Register />);
    const errorButton = component.find(ErrorButton);
    errorButton.simulate('click');
  });

  test('should click on errorButton elements with errors list', () => {
    useContext.mockImplementationOnce(
      jest.fn(() => ({
        error: ['a', 'b'],
        clearErrors: jest.fn()
      }))
    );

    const component = shallow(<Register />);
    const errorButton = component.find(ErrorButton).at(0);
    errorButton.simulate('click');
  });
});
