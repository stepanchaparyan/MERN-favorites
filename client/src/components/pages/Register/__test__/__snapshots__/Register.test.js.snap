// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register component test with Enzyme renders snapshot 1`] = `
<styled.div>
  <styled.div>
    Sign Up
  </styled.div>
  <styled.form
    onSubmit={[Function]}
  >
    <styled.input
      name="name"
      onChange={[Function]}
      placeholder="Name"
      required={true}
      type="text"
      value=""
    />
    <styled.input
      name="email"
      onChange={[Function]}
      placeholder="Email"
      required={true}
      type="email"
      value=""
    />
    <styled.input
      name="password"
      onChange={[Function]}
      placeholder="Password"
      required={true}
      type="password"
      value=""
    />
    <styled.input
      name="password2"
      onChange={[Function]}
      placeholder="Confirm Password"
      required={true}
      type="password"
      value=""
    />
    <styled.button>
      Sign Up
    </styled.button>
  </styled.form>
  <styled.div>
    <styled.button
      onClick={[Function]}
      type="button"
    >
      a
    </styled.button>
  </styled.div>
  <styled.div>
    Already have an accout?
     Â 
    <Link
      to="/login"
    >
      Sign In
    </Link>
  </styled.div>
</styled.div>
`;
